subsys/bluetooth/host/hci_core.c

```diff
@@ -605,6 +605,19 @@ static void hci_num_completed_packets(struct net_buf *buf)
 			node = sys_slist_get(&conn->tx_pending);
 
 			if (!node) {
+				/* The datapath on CIS links is not torn down on disconnects.
+				 * This means that there might be completed packets events
+				 * from the controller after the diesconnect.
+				 * Ignore them as the buffers are freed already.
+				 */
+				if (conn->type == BT_CONN_TYPE_ISO &&
+				    conn->state == BT_CONN_DISCONNECT_COMPLETE) {
```

Comment:
Should we allow the connection to be in the state BT_CONN_DISCONNECTED as well?

Reply:
Sjanc is correct. A controller may not send any events for a disconnected connection handle. 

That should, however, already be caught by https://github.com/zephyrproject-rtos/zephyr/pull/81383/files#diff-774b3de86adb275da54becf2a0958a2d4dae03d63e47840037ecc575e9d658b4R591-R595

Reply:
> On the CIS central, the host is allowed to continue sending data on a disconnected CIS.

Can you elaborate on that? 
It is true that after a disconnection, the ISO data path on the central remains active (for whatever reason), but does that indicate that the host can actually send on it? 

The controller is regardless not allowed to send HCI_Number_Of_Completed_Packets for any connection handle that is terminated, which I believe includes the CIS on the central. So whether or not the host sends on a disconnected connection handle for a CIS (which really doesn't make sense), the controller shall not send HCI_Number_Of_Completed_Packets to the host for that connection handle. 

This behavior is also described by BT Core 6.0, Vol 4, Part E, 4.3:
> When the Host receives an HCI_Disconnection_Complete event, **the Host shall assume
that all unacknowledged HCI Data packets that have been sent to the Controller
for the returned Handle have been flushed, and that the corresponding data buffers
have been freed**. A Controller does not have to notify the Host about this in an
HCI_Number_Of_Completed_Packets or an HCI_Number_Of_Completed_Data_Blocks
event before the disconnection event.

My opinion is that a host sending data on a disconnected connection handle, regardless of whether it's a CIS on a central, shall be ignored by the controller.

Reply:
I will fix this in the controller or otherwise make sure that there is a spec errata that clears things up first.